File: utm-seclib-go/pkg/ussp-pki/pki-keys.go
package ussp_pki

import (
	"crypto"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"fmt"

	"utm-pki/pkg/utils"
)

var ELLIPTIC_CURVE elliptic.Curve = elliptic.P256()

const DEFAULT_RSA_KEY_SIZE int = 4096

var randomSeedBytes []byte

type RsaOptions struct {
	KeySize int
}

func init() {
	randomSeedBytes = make([]byte, 32)
	_, err := rand.Read(randomSeedBytes)
	if err != nil {
		panic("cannot seed math/rand package with cryptographically secure random number generator")
	}
}

func CreateEcdsaPrivateKey() (crypto.Signer, error) {

	return ecdsa.GenerateKey(ELLIPTIC_CURVE, rand.Reader)
}

func CreateEd25519PrivateKey() (crypto.Signer, error) {

	key := ed25519.NewKeyFromSeed(randomSeedBytes)
	if key == nil {
		return nil, fmt.Errorf("failed to generate private key")
	}

	return &key, nil
}

func CreateRSAPSSPrivateKey(rsaOptions *RsaOptions) (crypto.Signer, error) {

	if rsaOptions == nil {
		rsaOptions = &RsaOptions{KeySize: DEFAULT_RSA_KEY_SIZE}
	}

	privateKey, err := rsa.GenerateKey(rand.Reader, rsaOptions.KeySize)

	if err != nil {
		return nil, fmt.Errorf("failed to generate private key: %v", err)
	}

	return privateKey, nil
}

func GetPrivateKeyAsPemBytes(privateKey crypto.Signer) ([]byte, error) {
	var encodedBytes []byte
	if privateKey == nil {
		return nil, fmt.Errorf("private key is nil")
	}

	if !IsSupportedKey(privateKey) {
		return nil, fmt.Errorf("private key not supported")
	}

	keyLabel, err := GetKeyTypeLabel(privateKey)
	if err != nil {
		return nil, err
	}

	rawBytes, err := ParseBytesFromPrivateKey(privateKey)
	if err != nil {
		return nil, err
	}

	if len(rawBytes) == 0 {
		return nil, fmt.Errorf("failed to get raw bytes from private key")
	}

	pemBlock := pem.Block{
		Type:  keyLabel + " PRIVATE KEY",
		Bytes: rawBytes,
	}

	encodedBytes = pem.EncodeToMemory(&pemBlock)

	return encodedBytes, nil
}

func GetKeyTypeLabel(key crypto.Signer) (string, error) {
	if key == nil {
		return "", fmt.Errorf("key is nil")
	}

	switch key.(type) {
	case *rsa.PrivateKey:
		return "RSA", nil
	case *ecdsa.PrivateKey:
		return "ECDSA", nil
	case *ed25519.PrivateKey:
		return "Ed25519", nil
	default:
		return "", fmt.Errorf("unsupported key type")
	}
}

func ParseBytesFromPrivateKey(privateKey crypto.Signer) ([]byte, error) {
	if privateKey == nil {
		return nil, fmt.Errorf("private key is nil")
	}

	if !IsSupportedKey(privateKey) {
		return nil, fmt.Errorf("private key not supported")
	}

	switch key := privateKey.(type) {
	case *rsa.PrivateKey:
		return x509.MarshalPKCS1PrivateKey(key), nil
	case *ecdsa.PrivateKey:
		return x509.MarshalECPrivateKey(key)
	case *ed25519.PrivateKey:
		return key.Seed(), nil
	default:
		return nil, fmt.Errorf("unknown key type")
	}
}

func SavePrivateKeyToFile(privateKey crypto.Signer, filePath string) (byteCount int, err error) {
	err = validateKeyFileParameters(privateKey, filePath)
	if err != nil {
		return 0, err
	}

	_, errorNilIfFileExists := LoadPrivateKeyFromFile(filePath)
	if errorNilIfFileExists == nil {
		return 0, fmt.Errorf("file already exists")
	}

	pemBytes, err := GetPrivateKeyAsPemBytes(privateKey)
	if err != nil || len(pemBytes) < 1 {
		return 0, err
	}

	return utils.SaveBytesToFile(pemBytes, filePath)
}

func validateKeyFileParameters(key crypto.Signer, path string) error {
	if key == nil {
		return fmt.Errorf("private key is nil")
	}

	if !IsSupportedKey(key) {
		return fmt.Errorf("private key not supported")
	}

	if path == "" {
		return fmt.Errorf("path is empty")
	}

	return nil
}

func IsSupportedKey(key crypto.Signer) bool {
	switch key.(type) {
	case *rsa.PrivateKey:
		return true
	case *ecdsa.PrivateKey:
		return true
	case *ed25519.PrivateKey:
		return true
	default:
		return false
	}
}

File: utm-seclib-node/src/ussp-pki/pki-keys.ts
import crypto, { KeyObject, generateKeyPair } from 'crypto';
import fs from 'fs';
import { promisify } from 'util';

const generateKeyPairAsync = promisify(generateKeyPair);

const ELLIPTIC_CURVE = 'prime256v1';
const DEFAULT_RSA_KEY_SIZE = 4096;

export interface RsaOptions {
    keySize: number;
}

const randomSeedBytes = crypto.randomBytes(32);

export async function CreateEcdsaPrivateKey(): Promise<KeyObject> {
    try {
        const { privateKey } = await generateKeyPairAsync('ec', {
            namedCurve: ELLIPTIC_CURVE,
        });
        return privateKey;
    } catch (error) {
        const message = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Failed to generate ECDSA private key: ${message}`);
    }
}

export async function CreateEd25519PrivateKey(): Promise<KeyObject> {
    try {
        const { privateKey } = await generateKeyPairAsync('ed25519', {
            seedLength: 32,
            seed: randomSeedBytes,
        });
        return privateKey;
    } catch (error) {
        const message = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Failed to generate Ed25519 private key: ${message}`);
    }
}

export async function CreateRSAPSSPrivateKey(rsaOptions?: RsaOptions): Promise<KeyObject> {
    const keySize = rsaOptions?.keySize || DEFAULT_RSA_KEY_SIZE;
    try {
        const { privateKey } = await generateKeyPairAsync('rsa', {
            modulusLength: keySize,
            publicExponent: 0x10001,
        });
        return privateKey;
    } catch (error) {
        const message = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Failed to generate RSA-PSS private key: ${message}`);
    }
}

export function GetPrivateKeyAsPemBytes(privateKey: KeyObject): Buffer {
    if (privateKey.type !== 'private') {
        throw new Error('Not a private key');
    }

    if (!IsSupportedKey(privateKey)) {
        throw new Error('Private key not supported');
    }

    try {
        return Buffer.from(privateKey.export({
            type: 'pkcs8',
            format: 'pem'
        }));
    } catch (error) {
        const message = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Failed to export private key as PEM: ${message}`);
    }
}

export function GetKeyTypeLabel(key: KeyObject): string {
    if (!key) {
        throw new Error('Key is null');
    }

    switch (key.asymmetricKeyType) {
        case 'rsa-pss':
        case 'rsa':
            return 'RSA';
        case 'ec':
            return 'ECDSA';
        case 'ed25519':
            return 'Ed25519';
        default:
            throw new Error('Unsupported key type');
    }
}

export async function SavePrivateKeyToFile(privateKey: KeyObject, filePath: string): Promise<number> {
    await validateKeyFileParameters(privateKey, filePath);

    try {
        // Check if file exists
        if (fs.existsSync(filePath)) {
            throw new Error('File already exists');
        }

        const pemBytes = GetPrivateKeyAsPemBytes(privateKey);
        await fs.promises.writeFile(filePath, pemBytes);
        return pemBytes.length;
    } catch (error) {
        const message = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Failed to save private key to file: ${message}`);
    }
}

export async function LoadPrivateKeyFromFile(filePath: string): Promise<KeyObject> {
    try {
        const pemBytes = await fs.promises.readFile(filePath);
        return crypto.createPrivateKey(pemBytes);
    } catch (error) {
        const message = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Failed to load private key from file: ${message}`);
    }
}

async function validateKeyFileParameters(key: KeyObject, path: string): Promise<void> {
    if (!key) {
        throw new Error('Private key is null');
    }

    if (!IsSupportedKey(key)) {
        throw new Error('Private key not supported');
    }

    if (!path) {
        throw new Error('Path is empty');
    }
}

export function IsSupportedKey(key: KeyObject): boolean {
    if (!key) return false;

    const supportedTypes = ['rsa', 'rsa-pss', 'ec', 'ed25519'];
    return supportedTypes.includes(key.asymmetricKeyType || '');
}